proposals/ytt/005-semver/README.md

```diff
@@ -0,0 +1,217 @@
+
+A semantic version is a string, e.g. `3.2.12`, `0.23.0-beta`, `1.99.2-next+build.5`, which contains structured
+information about the version. Configuration authors may want
+to read a version's constituents, mutate them and compare different versions. For example, when authoring `ytt`
+templates:
+
+* I want to assert that a given string is a well-formed semver.
+* I want to change a semver's representation from string another to data structure, e.g. object or array.
+* I want to compare semvers, e.g. expect a given version to be larger than a known version.
+* I want to (de,in)crement the major, minor and/or patch version of a semver.
+* I want to add, change or remove a semver's prerelease identifier or metadata.
+
+Starlark's standard library provides the means to achieve this. However, it sets a high bar and burdens configuration
+authors to write boilerplate code for common, industry standard operations.
+
+## Terminology / Concepts
+
+A semantic version encodes different properties about the version in a string of the
+form `<major>.<minor>.<patch>(-<pre-release>)(+<metadata>)`.
+
+Semantic versions are strictly sortable.
+
+For its full specification refer to [Semantic versioning](https://semver.org/).
+
+## Proposal
+
+The proposal is to include a built-in `@ytt:semver` library. Configuration authors can then load it and programmatically
+work with strings representing semvers.
+
+The semver library contains means for turning a semver string into a structured representation. This can result in an
+error in the case of a malformed semver. If the operation succeeds, it returns a `struct` representing the semver.
+
+Now, it is possible to read the _major_, _minor_, _patch_ versions as integers, as well as the _pre-release_
+version and _build metadata_ as a strings. This enables authors to change its representation at will.
+
+It is also possible to programmatically construct a semver `struct` by providing _major_, _minor_, _patch_ versions as
+required integers, as well as the _pre-release_ version and _build metadata_ as optional strings. This enables authors
+to construct semvers either from values known in advance or which are received in a representation other than a semver.
+
+Given two semver `structs`, they can be compared. This enables authors to test if a semver is large or small enough,
+or whether it is within a range.
+
+For its implementation, the suggestion is to use [blang/semver](https://github.com/blang/semver). It is a
+well-established module, contains an API to address all conceivable case and is already used by Carvel's `vendir` for
+its `tools sort-semver` subcommand.
+
+### Goals and Non-goals
+
+* Enable configuration authors to programmatically work with semvers out-of-the-box.
+
+* Education around and proliferation of the semver scheme is not considered to be `ytt`'s responsibility.
+
+### Specification / Use Cases
+
```

Comment:
TODO: fold in any new use case / needs from [Package Version Constraints in kapp-controller](https://hackmd.io/6FQhzRIPR_uceqtTqSIXaA)